server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -123,40 +126,90 @@ protected TransportNodesAction(
 
     @Override
     protected void doExecute(Task task, NodesRequest request, ActionListener<NodesResponse> listener) {
-        new AsyncAction(task, request, listener).start();
-    }
+        if (request.concreteNodes() == null) {
+            resolveRequest(request, clusterService.state());
+            assert request.concreteNodes() != null;
+        }
 
-    /**
-     * Map the responses into {@code nodeResponseClass} responses and {@link FailedNodeException}s, convert to a {@link NodesResponse} and
-     * pass it to the listener. Fails the listener with a {@link NullPointerException} if {@code nodesResponses} is null.
-     *
-     * @param request The associated request.
-     * @param nodeResponseTracker All node-level responses collected so far
-     * @throws NodeResponseTracker.DiscardedResponsesException if {@code nodeResponseTracker} has already discarded the intermediate results
-     * @see #newResponseAsync(Task, BaseNodesRequest, List, List, ActionListener)
-     */
-    // exposed for tests
-    void newResponse(Task task, NodesRequest request, NodeResponseTracker nodeResponseTracker, ActionListener<NodesResponse> listener)
-        throws NodeResponseTracker.DiscardedResponsesException {
+        final var mutex = new Object();
+        final var responses = new ArrayList<NodeResponse>(request.concreteNodes().length);
+        final var exceptions = new ArrayList<FailedNodeException>(0);
 
-        if (nodeResponseTracker == null) {
-            listener.onFailure(new NullPointerException("nodesResponses"));
-            return;
+        final var resultListener = new ListenableFuture<NodesResponse>();
+        final var resultListenerCompleter = new RunOnce(() -> {
+            if (task instanceof CancellableTask cancellableTask) {
+                if (cancellableTask.notifyIfCancelled(resultListener)) {
+                    return;
+                }
+            }
+            // ref releases all happen-before here so no need to be synchronized
+            threadPool.executor(finalExecutor)
+                .execute(ActionRunnable.wrap(resultListener, l -> newResponseAsync(task, request, responses, exceptions, l)));
+        });
+
+        final var nodeFailureListeners = new ListenableFuture<NodeResponse>();
```

Comment:
If I am not mistaken this is mainly used to notify about the task being cancelled right? What do you think about naming it `nodeCancellationListener` or something similar. My reasoning is that we use quite a few listeners here and it was a bit hard for me to follow.

Reply:
Yes that's right this is just used to complete the outstanding node listeners (releasing refs to `responses` and `exceptions`) on cancellation. Naming things is hard :)

Reply:
Indeed, it is :)